File: Data Structures & Algorithms/longest-turbulent-subarray/submission-1.py
class Solution:
    def maxTurbulenceSize(self, arr: List[int]) -> int:
        left, right = 0, 0
        cur_sign = 0  # -1: less, 1: greater
        ans = 0
        if len(arr) < 2:
            return len(arr)
        
        for right in range(1, len(arr)):
            if arr[right] == arr[right - 1]:
                left = right
                cur_sign = 0
            elif cur_sign == -1 and arr[right - 1] < arr[right]:
                left = right - 1
            elif cur_sign == 1 and arr[right - 1] > arr[right]:
                left = right - 1
            ans = max(right - left + 1, ans)
            cur_sign = -1 if arr[right - 1] < arr[right] else 1

        return ans
        

File: Data Structures & Algorithms/longest-substring-without-duplicates/submission-1.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        // sliding window
        if (s.empty())
            return 0;
        
        // last seen index
        std::array<int, 128> ls;
        ls.fill(-1);
        
        const int N = s.size();
        int left = 0;
        int ans = 0;
        
        for (int right = 0; right < N; ++right) {
            if (ls[(unsigned char)s[right]] >= left) {
                // found a duplicate char
                left = ls[s[right]] + 1;
                // continue;
            }

            ans = std::max(ans, right - left + 1);
            ls[(unsigned char)s[right]] = right;
        }

        return ans;
    }
};
